feat(integration-test-runner): Add Cypress

michaelw85 opened the pull request:
Introduce Cypress for e2e testing.

Feature request: #943

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aurelia/cli?pullRequest=1029) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aurelia/cli?pullRequest=1029) before we can accept your contribution.<br/><hr/>**Michael Withagen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aurelia/cli?pullRequest=1029) it.</sub>

bigopon commented:
ping @JeroenVinke @huochunpeng @EisenbergEffect

huochunpeng commented:
Probably need to merge #993 first, that one changed the flow quite a bit.

EisenbergEffect commented:
@michaelw85 Thanks for working on this. It's an awesome contribution! We had a recent set of work around project scaffolding that we were doing. Completely my fault, it dropped off my radar and I didn't get it merged earlier. I just merged it now. Unfortunately, that creates some conflicts with this PR. Would you mind taking a look at those and resolving them. I apologize for the inconvenience.

michaelw85 commented:
No problem, I already anticipated (some) rework. I will try to look asap.

michaelw85 commented:
@EisenbergEffect Rebased & resolved a minor collision, re-tested creating a new project and still works as expected. When you have time could you please review.

huochunpeng commented:
Looks like you have some duplicated commits after rebase.

Could you merge all commits into one, and follow the [commit message convention](https://github.com/DurandalProject/about/blob/master/CONTRIBUTING.md#commit)?

It's probably same as your PR title and message.
```
feat(test): Add Cypress integration test runner

Introduce Cypress for e2e testing.

closes #943
```

michaelw85 commented:
Ok I will try to do that have to figure out how but Google can surely assist me with that

michaelw85 commented:
@huochunpeng Done, merged all commits into one and updated the commit message.

I do wonder if this was really necessary though, when merging to master it is possible to squash all changes producing the same result. 
For future contributions should I follow the commit guidelines per commit on a branch (I noticed this with other branches/pr's) or would it suffice for the PR itself?

huochunpeng commented:
Normally I am not so picky to ask that. I asked because I saw some duplicated commits likely due to some strange rebase/merge.

huochunpeng commented:
I will help to test this feature tomorrow.

huochunpeng commented:
Btw, I think you need to rewrite the commit message to follow the convention, so our release task can write change log properly from git history.

michaelw85 commented:
😭 Just force pushed the merge again.... sorry I thought I would look into github for desktop and it synced the branch and reverted the merge.

huochunpeng commented on lib/commands/new/buildsystems/general/integration-test-runner/cypress.js:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  project.addToTasks(
+    ProjectItem.resource('tsconfig.json', 'content/tsconfig.template.json')
```
Is this a mistake?

michaelw85 replied:
I'm assuming you are referring to the tsconfig always being added? I should have made this conditional.

huochunpeng replied:
The top level tsconfig is handled by transpilers/typescript. I don't think you want copy it again here (even with condition).

michaelw85 replied:
I see, you are right. Removed handling the tsconfig at this level.

huochunpeng commented on lib/package-managers/npm.js:
```diff
@@ -8,34 +8,18 @@ exports.NPM = class {
     this.executableName = 'npm';
   }
 
-  install(packages, workingDirectory) {
```
Please leave the npm.js refactor to a separated PR. I guess you are trying to fix the hidden output from npm.
But this destroyed the yarn.js which extends npm.js, you need to test both npm and yarn.

michaelw85 replied:
No this was done because npm install downloads cypress and extracts the archive but in the old setup each update of progress outputs as a single line which causes maximum buffer exception which breaks the install and you have to manually remove the package and run npm install. But yes as a side effect this also fixes the hidden output.

huochunpeng replied:
Ok, then you have to make sure yarn works. The previous installCommand method is to give yarn a chance to overwrite the command format. You could just get the old installCommand back.

npm
```sh
npm i # for install all
npm i a b c # for install packages
```

yarn
```sh
yarn # for install all
yarn add a b c # for install packages
```

michaelw85 replied:
But my change was only applied for npm, yarn.js is still intact with the original install method. I assumed changing the npm.js would not affect yarn in any way.

huochunpeng replied:
```js
exports.Yarn = class extends npm
```

That's where it broke, your new implementation will execute `yarn i a b c`.

michaelw85 replied:
ugh that's very ugly and I wish this was written using typescript 😭 
I will have a look how to fix this.

michaelw85 replied:
Not the prettiest implementation I could have done, moved the original setup into the Yarn class. Not allot being shared anymore. I think I should have a look in a separate PR and see if I can refactor this into a nicer setup.

huochunpeng commented on lib/package-managers/npm.js:
```diff
@@ -8,34 +8,18 @@ exports.NPM = class {
-  install(packages, workingDirectory) {
-    if (!workingDirectory) workingDirectory = process.cwd();
-
+  install(packages = [], workingDirectory = process.cwd()) {
```
The reason we did not give workingDirectory a default value is because cli allows to issue the command from a sub folder (for example `your-app-folder/src`).

The cli finds out correct working directory which **may be different from process.cwd()**. So this `install()` method is always called with an explicit workingDirectory internally. You don't need a default value here, and `process.cwd()` could be wrong default.

michaelw85 replied:
Ok I will remove the default value but this will functionally be different from the previous original implementation where the if statement did the same fallback if no working dir was passed to this method.

huochunpeng replied:
Sorry ignore my previous message, I forgot how the code works :(

huochunpeng commented:
Sorry, I saw you force-pushed before seeing my confession on mistake. To be clear `workingDirectory = process.cwd()` is correct. Sorry about that.

michaelw85 commented:
> Sorry, I saw you force-pushed before seeing my confession on mistake. To be clear `workingDirectory = process.cwd()` is correct. Sorry about that.

I just reverted it in combination with an attempt to fix yarn. I do not have yarn setup so not sure how to test/force it to be used.

michaelw85 commented:
Just tested a default project with Yarn, deps are installed as expected 😄

huochunpeng commented on lib/package-managers/npm.js:
```diff
@@ -8,34 +8,18 @@ exports.NPM = class {
-      installProcess.stdout.on('data', data => console.log(data));
+      childProcess.spawn(
+        this.getExecutablePath(workingDirectory),
+        ['i', ...packages],
```
I think you still can normalise this line to a method, so yarn can override the args.

michaelw85 replied:
Ok updated the code, changed the abbreviation to 'install' which works for Yarn. Checked the code and seems like `installCommand` is not used at all so I decided to also remove this from the Yarn class.

michaelw85 commented on lib/package-managers/npm.js:
```diff
@@ -8,34 +8,18 @@ exports.NPM = class {
-      installProcess.stdout.on('data', data => console.log(data));
+      childProcess.spawn(
+        this.getExecutablePath(workingDirectory),
+        ['install', ...packages],
```
I added a base package manager class allowing a specific overwrite for Yarn add. Tested the auto install feature and works like a charm (awesome feature btw!).

huochunpeng commented on lib/resources/content/cypress.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "baseUrl": "http://localhost:8080",
```
8080 is the default port for our webpack setup. But cli bundler uses default port 9000. Need to separate template for webpack and cli bundler.

huochunpeng replied:
FYI, protractor conf is js file, it reads the port from aurelia.json.
```js
const aureliaConfig = require('../aurelia_project/aurelia.json');
const port = aureliaConfig.platform.port;
```

michaelw85 replied:
Good point, I completely overlooked this configuration. Will change asap

michaelw85 replied:
Fixed, moved config into JS file and use via aurelia task. Cypress does not support a js config out of the box but I am able to pass in the configuration when running Cypress.

huochunpeng commented on lib/resources/test/cypress/integration/app.e2e.js:
```diff
@@ -0,0 +1,13 @@
+  });
+
+  it('should load the page and display the initial page title', () => {
+    cy.title().should('equal', 'Aurelia Navigation Skeleton');
```
A small problem. There is newly added scaffold.
```
Which features do you want scaffolded into your project?

1. None
   Just give me an empty project.
2. Navigation
   Add a router and some sample routes. (Also adds Bootstrap 3 and Font Awesome.)
```

When user picks "navigation" scaffold, the e2e needs to be different. Again, you can look into existing protractor setup.

michaelw85 replied:
Ok will do 👍

huochunpeng commented:
The cypress is working well for me.

huochunpeng commented:
I don't have the privilege to edit. Can you fix your PR title from `feat:(integration-test-runner):` to `feat(integration-test-runner):` please? Thx!

huochunpeng commented on lib/resources/test/cypress/integration/app.e2e.ts:
```diff
@@ -0,0 +1,13 @@
+  });
+
+  it('should load the page and display the initial page title', () => {
+    cy.title().should('equal', 'Aurelia Navigation Skeleton');
```
Probably just remove the title test.

I just found out not all our skeleton has the same page title, which is largely our fault. :-)

michaelw85 replied:
Removed the title test from default setup, left it in place for the navigation variant.

huochunpeng commented on lib/resources/test/cypress/plugins/index.ts:
```diff
@@ -0,0 +1,34 @@
+// This function is called when a project is opened or re-opened (e.g. due to
+// the project's config changing)
+
+const wp = require('@cypress/webpack-preprocessor')
```
This plugin supports TypeScript, but it assumes webpack in place.

It fails with cli-bundler + systemjs + typescript setup where webpack is not available.

However, after I manually install webpack for the systemjs/ts setup. Cypress is able to run.

michaelw85 replied:
Looking into it. 
Possible quick/easy solution would be to add webpack to as a dep.
Investigating if I could use the cli-bundler as a preprocessor.

huochunpeng replied:
Not cli bundler, but directly use typescript compiler. It seems cypress does not have hard dep on webpack or browserify.

michaelw85 replied:
Yes that's a good idea, I was just reading the typescript compiler api examples. I also noticed cypress does not have those deps although somewhere in the docs it states it is shipped with browserify...

michaelw85 replied:
Added webpack & ts-loader as dep to get it working for now.

huochunpeng commented:
@bahmutov can we borrow your expertise here?

aurelia-cli has an option to build app using a built-in bundler (AMD based), which is not webpack.

The TypeScript support of @michaelw85's sollution is a reference cypress plugin using `@cypress/webpack-preprocessor`. However, any app based on aurelia-cli built-in bundler doesn't have webpack. The cypress plugin for TS support doesn't work without webpack.

Is it easy to have a cypress plugin for TS that is purely on top of TypeScript compiler itself?

michaelw85 commented:
@huochunpeng I've spend a couple of hours trying to get typescript compiler working instead of using the webpack based plugin. I tried to use gulp which is already available in the cli bundler + systemjs + TS setup but I only got it working to a certain level. I can't get it to properly run with all of the same features the webpack plugin offers.

I think it's much simpler and more reliable to include webpack as a dev dep for the cli/systemjs/ts setup and make use of the official Cypress webpack/ts plugin to get the TS setup working. I personally don't see a problem in using webpack just for Cypress even though you do not choose to use it for Aurelia itself. I do think it would be nice to not add more deps to the cli/systemjs/ts setup but the amount of effort vs the benefit does not seem to be worth writing a custom cypress ts preprocessor.

If this is not an option I would prefer removing cypress in case someone chooses cli bundler + systemjs + ts and add a feature request to the Cypress backlog for a systemjs/gulp preprocessor.

huochunpeng commented:
Thx for the effort. I too think adding webpack to deps is easier solution.

zewa666 commented:
Even if we go the route of adding webpack as a dependency, I'd argue it would be good to raise this concern with the Cypress guys. For me it just feels plain wrong having to have Webpack just for the sake of getting Cypress up and running and I'm sure the guys there do feel the same.

So maybe along @bahmutov giving his thought here we should also create an issue in the cypress repo.

michaelw85 commented:
@zewa666 It's not needed to get Cypress up and running. It is needed to be able to use Typescript for writing tests in Cypress. Cypress offers 2 preprocessors plugins; one based on webpack (which I used in my setup since I prefer webpack) the other option is Browserify.

bahmutov commented:
@michaelw85 is right - you only need custom Webpack plugin to point Cypress at your own webpack config file to compile the spec files (and anything they import) using your current webpack config. This also unfortunately includes TypeScript transpile. If you are ok with coding specs in JavaScript or CoffeeScript nothing is necessary.

PS: really excited this PR is progressing

huochunpeng commented:
@bahmutov thanks for the clarification. Why not provide a reference TS plugin plainly on top of TypeScript compiler? I can make an issue of feature request to cypress if this makes sense.

bahmutov commented:
I would appreciate if you open an issue about, then we can discuss it with
Chris B

On Mon, Feb 11, 2019 at 3:49 PM huochunpeng <notifications@github.com>
wrote:

> @bahmutov <https://github.com/bahmutov> thanks for the clarification. Why
> not provide a reference TS plugin plainly on top of TypeScript compiler? I
> can make an issue of feature request to cypress if this makes sense.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aurelia/cli/pull/1029#issuecomment-462488597>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACHApna_s_HO9ACKjzMvOrMOdL6iWYMFks5vMddPgaJpZM4acaA5>
> .
>


-- 
Dr. Gleb Bahmutov, PhD

Schedule video chat / phone call / meeting with me via
https://calendly.com/bahmutov
gleb.bahmutov@gmail.com @bahmutov <https://twitter.com/@bahmutov>
https://glebbahmutov.com/ https://glebbahmutov.com/blog
https://github.com/bahmutov

huochunpeng commented:
@zewa666 can we good on bringing webpack for built-in bundler + TypeScript setup? Cypress is not going to have a new plugin anytime soon.

zewa666 commented:
Is that only added if you choose Cypress during project scaffolding? Or is it a default dependency even if you skip e2e?

michaelw85 commented:
> Is that only added if you choose Cypress during project scaffolding? Or is it a default dependency even if you skip e2e?

It is conditional for Cypress + Typescript:
```
  // lib/commands/new/buildsystems/general/integration-test-runner/cypress.js:9
  if (isUsingTypescript) {
    project.addToDevDependencies('@cypress/webpack-preprocessor');
    project.addToDevDependencies('webpack');
    project.addToDevDependencies('ts-loader');
  }
```

zewa666 commented:
@michaelw85 first let me start with saying a big thank you in the name of the team for taking the time to work into this topic. I really don't want to come over as the grumpy grandpa who sabotages your efforts just because he's nitpicky ;) But ... obviously ;), I feel that going forward and add webpack as a fixed dependency even if you're not using webpack as the bundler of choice, is a totally subpar solution. Given your project scaffold picked webpack why not this would be fine but for the other scenarios I see that more as a burden then benefit.

I dunno but I prefer my dependencies lean and clean and especially free of large bloated libs. Now bundlers are already bad but what is even worse are two of them ;)

So I don't want to be the blocking part here. If both @huochunpeng and @EisenbergEffect agree that it's good enough to go I'll definitely agree with them, but I personally would prefer to pursue the linked topic a bit further and see if can't get the Cypress guys to overthink their opinion. I think it's not only the best for us but also for their product to offer a fundamental basic solution for TS that is valid for all bundler scenarios.

michaelw85 commented:
@zewa666 I completely agree and I don't see this as nitpicking. 
My goal of course is to get this PR merged and prevent it from going stale and try to keep it as small as possible. I would like to find an acceptable state which we/I could improve on in the near future.

To be able to move forward I see a couple options: 
1) We keep it as is adding webpack+ts-loader as dep if @huochunpeng & @EisenbergEffect agree/accept.
2) I add the JS setup instead of TS for Cypress since we cannot properly support TS at this time (sub optimal solution imo, I would personally prefer adding a dep to enable TS than fall back to JS)
3) I remove Cypress as an option when the cli bundler was chosen in combination with TS.

In terms of effort and enrichment of the cli/aurelia I would prefer option 1. If we want to keep it a bit more clean and want to offer Cypress in any Aurelia combination I would go for option 2. Last but not least the cleanest solution option 3 we cannot properly support cli bundler+TS in combination with Cypress until a TS preprocessor is available for Cypress that does not rely on webpack/browserify. (I could try to continue my attempt to get it working using Gulp since its already a dep for us)

huochunpeng commented:
Engineering is a task to compromise between idea and reality. I am at the lazy side of engineer spectrum 😄 . This is practically good enough to me. 

Rob is out of internet, he will take some time to respond.

EisenbergEffect commented:
What if we go for option 2 but add descriptive information into the CLI menu option that states that it will only work with JS. That way, when someone is choosing their e2e solution, they have the option to use Cypress in all configurations, but are informed in their decision to use it vs. another solution if they care about their tests being in TypeScript or not. Thoughts?

huochunpeng commented:
@EisenbergEffect there is one downside of the all-js solution.

Majority of cli users are webpack users, if they use TS, we should provide TS tests files too. We don't want to confuse webpack users with a problem from non-webpack setup.

EisenbergEffect commented:
Makes sense. Can we provide a hybrid solution with TS for TS-Webpack and JS for everything else, but in the case of a TS setup without Webpack, provide option text that explains that tests will be in JS for the Cypress option?

huochunpeng commented:
It could be done.

cli use `'app.e2e.ext'` with `project.model.transpiler` to find the currrent file.
Now you need `usesWebpack ? project.model.transpiler : 'babel'`, bit hacky.

michaelw85 commented:
Instead of offering JS with a sidenote why not add JS and Cypress+TS with a note it will add webpack as a dep. This way we do not force webpack but do offer TS and the choice is up to user. This could be done for all paths so no complex branches

EisenbergEffect commented:
I'm not sure which way is better to be honest. Anyone else on this thread have an opinion?

michaelw85 commented:
My reason is because this discussion started that we would force a dep to a user that explicitly went with a setup that did not include it. If we make it the responsibility of the user to make a decision then we do not force anything and they should get the best possible fit for their needs/preferences.

zewa666 commented:
I think I start to tend towards opt 1 along an additional note as @michaelw85 suggested that if you use a TS setup and you haven't picked Webpack before.

Something along the lines that it's a current limitation and we need Webpack as a dependency in order to get going with TS, if the user does not want to do so, hint about the alternative with Protractor.

EisenbergEffect commented:
It sounds like we're leaning towards Option 1 but with an informational note that allows developers to make an informed choice WRT tradeoffs. I'm good with that. Any major disagreement here?

davismj commented:
Option (1) seems fine to me since you have to opt into it. It seems like it can't be helped for now without tremendous effort. If down the line Cypress supports more general loading options, perhaps we can revisit?

huochunpeng commented:
When cypress releases a tsc plugin in future. Option 1 is the easiest to upgrade.

Futuremore, any option has to face the reality of our workflow implementation. We really don’t want to make supre complex json config just for cypress.

EisenbergEffect commented:
Ok, let's go with option 1 😄 Please ping me and @huochunpeng one final time when it's ready for review and we'll get it in. Thanks everyone!

michaelw85 commented:
Ok will do! Hoping to be able to find some time tonight (8:30 AM here atm).

michaelw85 commented:
@huochunpeng @EisenbergEffect Added informative text to the Cypress option:

![capture](https://user-images.githubusercontent.com/3629094/52807153-04c33480-308b-11e9-9588-d50b023f2b7a.PNG)

EisenbergEffect commented:
Works for me.

EisenbergEffect commented:
Set for a final review then @michaelw85 ?

michaelw85 commented:
Yes I think so :)

huochunpeng commented:
Found one last issue :-(
Calling TypeScript expert to help.

Choose cli-bundler + TypeScript + karma(jasmine) + cypress(mocha), there is typing conflicts raised by TypeScript compiler because both jasmine and mocha defined global vars "describe/it/...".

```
/Users/huocp/playground/t/node_modules/@types/jasmine/index.d.ts(23,18): error TS2300: Duplicate identifier 'describe'.
/Users/huocp/playground/t/node_modules/@types/jasmine/index.d.ts(25,18): error TS2300: Duplicate identifier 'xdescribe'.
/Users/huocp/playground/t/node_modules/@types/jasmine/index.d.ts(34,18): error TS2300: Duplicate identifier 'it'.
/Users/huocp/playground/t/node_modules/@types/jasmine/index.d.ts(44,18): error TS2300: Duplicate identifier 'xit'.
/Users/huocp/playground/t/node_modules/@types/mocha/index.d.ts(36,13): error TS2300: Duplicate identifier 'describe'.
/Users/huocp/playground/t/node_modules/@types/mocha/index.d.ts(37,13): error TS2300: Duplicate identifier 'xdescribe'.
/Users/huocp/playground/t/node_modules/@types/mocha/index.d.ts(42,13): error TS2300: Duplicate identifier 'it'.
/Users/huocp/playground/t/node_modules/@types/mocha/index.d.ts(43,13): error TS2300: Duplicate identifier 'xit'.
```

huochunpeng commented:
One more bug. But in existing cli code.

Choose webpack + cypress, there is many missing devDependencies. Because both webpack and cypress try to do `project.addToDevDependencies('webpack',...)`.

The bug is here in `lib/commands/new/project-template.js`.
```js

function addDependencies(current, toAdd) {
  for (let i = 0, ii = toAdd.length; i < ii; ++i) {
    let name = toAdd[i];

    if (name in current) {
      return; // BUG, change to "continue;"
    }

    if (name.indexOf('@') <= 0) {
      current[name] = getSupportedVersion(name).replace('\"', '').replace('\"', '');
    } else {
      let split = name.split('@');
      current[split[0]] = split[1];
    }
  }
}
```

michaelw85 commented:
Thx for the info, I will try to look into the issues asap. Sorry for wasting your time.

huochunpeng commented:
> Sorry for wasting your time.

Don't say that pls.

EisenbergEffect commented:
@michaelw85 No need to apologize. We're hugely grateful for your work on this :)

zewa666 commented:
With regards to the types conflict there is no real workaround except making sure only one type definition is installed. The workaround via compilerOptions.types array would be to manually include all types you'd like to get resolved which is a super bummer.

huochunpeng commented:
@zewa666 in code pls :)

michaelw85 commented:
@huochunpeng Fixed the duplicate dependency issue by changing return to continue.
Also fixed the types issue by adding the types array as @zewa666 mentioned, I've only done this for CLI Bundler + TS + Karma + Cypress. I did some reading to see the possible impact of adding this setting and I think we should be fine. It disables automatic loading for types but this only impact globally declared modules that are not imported into file like jasmine and mocha.

I quote from the [TS documentation](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html):

> If types is specified, only packages listed will be included. For instance:

```js
{
   "compilerOptions": {
       "types" : ["node", "lodash", "express"]
   }
}
```

> This tsconfig.json file will only include ./node_modules/@types/node, ./node_modules/@types/lodash and ./node_modules/@types/express. Other packages under node_modules/@types/* will not be included.
> 
> A types package is a folder with a file called index.d.ts or a folder with a package.json that has a types field.
> 
> Specify "types": [] to disable automatic inclusion of @types packages.

This is the most interesting part:
> Keep in mind that automatic inclusion is only **important if you’re using files with global declarations** (as opposed to files declared as modules). **If you use an import "foo" statement, for instance, TypeScript may still look through node_modules** & node_modules/@types folders to find the foo package.

vscode validates this behavior for me and show the correct intellisense. I've also run Karma and Cypress test and started the application after adding the types array and it seems to work as expected.

huochunpeng commented:
Everything works fine except one more strange issue. Sorry for the disappointment 😆 

When I run `au cypress --run`, it gave error:
```
⋊> ~/p/wk au cypress --run
Local aurelia-cli v1.0.0-beta.13
Could not find any tests to run.

We looked but did not find a cypress.json file in this folder: /Users/huocp/playground/wk
```

But strangely, the problem goes away if I do a manual test with `au cypress` before trying `au cypress --run`. It must be an easy fix for you.

michaelw85 commented:
Hmm that could be related to config refactoring I did. Will have a look asap

michaelw85 commented:
@huochunpeng Tested and reproduced the issue. It seems that the cypress.json is required for the run command. When you use `cypress.open()` (this is done when you use `au cypress`) it generates the config file. I removed it in previous refactoring, replaced it with a JS file to be able to dynamically inject the port for the base url. JS format is not supported out of the box so I just passed the config into the node `cypress.open` and `.run` commands and figured it would be ok. I only tested the run command after the open so never ran into this scenario.

TL;DR Fixed it 😉

huochunpeng commented:
Thx, all work for me now.

@EisenbergEffect let's get this in.

EisenbergEffect commented:
Thank you everyone!
